Firestore/core/src/firebase/firestore/core/relation_filter.cc

```diff
@@ -41,48 +67,76 @@ const FieldPath& RelationFilter::field() const {
-    abort();
+    HARD_ASSERT(value_rhs_.type() == FieldValue::Type::Reference,
+                "Comparing on key, but filter value not a Reference.");
+    HARD_ASSERT(op_ != Filter::Operator::ArrayContains,
+                "arrayContains queries don't make sense on document keys.");
+    const auto& ref = value_rhs_.reference_value();
+    ComparisonResult comparison = doc.key().CompareTo(ref.key());
+    return MatchesComparison(comparison);
   } else {
     absl::optional<FieldValue> doc_field_value = doc.field(field_);
     return doc_field_value && MatchesValue(doc_field_value.value());
   }
 }
 
-bool RelationFilter::MatchesValue(const FieldValue& other) const {
+bool RelationFilter::MatchesValue(const FieldValue& lhs) const {
   if (op_ == Filter::Operator::ArrayContains) {
-    if (other.type() != FieldValue::Type::Array) return false;
+    if (lhs.type() != FieldValue::Type::Array) return false;
 
-    const std::vector<FieldValue>& contents = other.array_value();
+    const auto& contents = lhs.array_value();
     return absl::c_linear_search(contents, value_rhs_);
   } else {
     // Only compare types with matching backend order (such as double and int).
-    return FieldValue::Comparable(other.type(), value_rhs_.type()) &&
-           MatchesComparison(other);
+    return FieldValue::Comparable(lhs.type(), value_rhs_.type()) &&
+           MatchesComparison(lhs.CompareTo(value_rhs_));
   }
 }
 
-bool RelationFilter::MatchesComparison(const FieldValue& other) const {
+bool RelationFilter::MatchesComparison(ComparisonResult comparison) const {
   switch (op_) {
     case Operator::LessThan:
-      return other < value_rhs_;
+      return comparison == ComparisonResult::Ascending;
     case Operator::LessThanOrEqual:
-      return other <= value_rhs_;
+      return comparison == ComparisonResult::Ascending ||
+             comparison == ComparisonResult::Same;
     case Operator::Equal:
-      return other == value_rhs_;
-    case Operator::GreaterThan:
-      return other > value_rhs_;
+      return comparison == ComparisonResult::Same;
     case Operator::GreaterThanOrEqual:
-      return other >= value_rhs_;
+      return comparison == ComparisonResult::Descending ||
+             comparison == ComparisonResult::Same;
+    case Operator::GreaterThan:
+      return comparison == ComparisonResult::Descending;
     case Operator::ArrayContains:
       HARD_FAIL("Should have been handled in MatchesValue()");
   }
   UNREACHABLE();
 }
 
 std::string RelationFilter::CanonicalId() const {
-  // TODO(rsgowman): Port this
-  abort();
+  return absl::StrCat(field_.CanonicalString(), Describe(op_),
+                      value_rhs_.ToString());
+}
+
+std::string RelationFilter::ToString() const {
+  return util::StringFormat("%s %s %s", field_.CanonicalString(), Describe(op_),
+                            value_rhs_.ToString());
+}
+
+size_t RelationFilter::Hash() const {
+  return util::Hash(field_, static_cast<int>(op_), value_rhs_);
```

Comment:
Very optional: perhaps we could add an overload of `Hash` that is SFINAE'd upon `std::is_enum` to avoid these casts in the future.

Reply:
I'll leave this for another day.